### PR TITLE
Truncate error stack trace.

### PR DESCRIPTION
**Pull request by wting**
We currently log the entire call stack, which sometimes exceeds our log line
limits (4096 bytes). By truncating to the first 2KB, it allows us to maintain
JSON parseable log lines while still supporting source maps for the ~16 most
recent function calls of the call stack.

:eyeglasses: @schwers @uzi 

<details><summary>example error stack overflow</summary><p>

```json
Oct 31 14:55:39 mweb2x-app-9aa979aa [mweb2x] 
{
  "env":"CLIENT"
"userAgent":"Mozilla/5.0 (iPhone; CPU iPhone OS 9_2 like Mac OS X) AppleWebKit/601.1.46 (KHTML like Gecko) Version/9.0 Mobile/13C75 Safari/601.1"
"isAPIFailure":false
"message":"Error: SecurityError: DOM Exception 18"
"requestUrl":"https://www.google.com/"
"reduxInfo":
  "Redux Action Stack: \"SAVE_SCROLL_POSITION\"
  \"FETCHING_RECOMMENDED_SUBREDDITS\"
  \"RECEIVED_COMMENTS_PAGE\"
  \"PLATFORM__SET_STATUS\"
  {\"type\":\"PLATFORM__SET_PAGE\"
  \"payload\":{\"url\":\"https://www.google.com/\"
  \"urlParams\":{\"0\":\"https://www.google.com/\"}
  \"queryParams\":{}
  \"hashParams\":{}
  \"referrer\":\"\"}}
  {\"type\":\"PLATFORM__NAVIGATE_TO_URL\"
  \"payload\":{\"method\":\"get\"
  \"pathName\":\"https://www.google.com/\"
  \"queryParams\":{}
  \"hashParams\":{}
  \"bodyParams\":{}
  \"referrer\":\"\"}}
  \"RECEIVED_RECOMMENDED_SUBREDDITS\"
  \"RECEIVED_SUBREDDIT\"
  \"SET_TITLE\"
  \"PLATFORM__SET_STATUS\""
"stack":
  "pushState@[native code]
  value@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:5:7323
  _performComponentUpdate@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:30:3452
  updateComponent@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:30:2837
  receiveComponent@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:30:2114
  receiveComponent@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:9:31234
  _updateRenderedComponent@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:30:3791
  _performComponentUpdate@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:30:3560
  updateComponent@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:30:2837
  performUpdateIfNecessary@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:30:2334
  performUpdateIfNecessary@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:9:31396
  u@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:7:19738
  perform@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:14:23638
  perform@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:14:23638
  perform@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:7:20737
  P@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:7:20909
  close@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:7:20288
  closeAll@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:14:24275
  perform@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:14:23725
  perform@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:7:20737
  P@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:7:20909
  [native code]
  closeAll@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:14:24275
  perform@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:14:23725
  batchedUpdates@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:30:27819
  s@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:7:19950
  r@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:18:26506
  enqueueSetState@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:18:27496
  setState@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:18:11377
  handleChange@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:29:14456
  [native code]
  l@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:19:9390
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:16:2909
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:16:4967
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:13:26199
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:37:17274
  dispatch@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:31:27623
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:15:11798
  i@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:31:30873
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:32:296
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:31:31055
  r@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:15:10903
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:15:11051
  w@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:23:8044
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:15:10858
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:15:12149
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:16:4757
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:13:26199
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:37:17274
  dispatch@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:31:27623
  k@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:20:29789
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:20:29251
  i@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:31:30873
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:32:296
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:31:31055
  r@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:20:26765
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:20:26866
  a@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:23:6917
  https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:23:7039
  c@https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:14:108"
```

</p></details>

<details><summary>example truncated error stack that still works with source maps</summary><p>

```json
{
    "env": "CLIENT",
    "isAPIFailure": true,
    "message": "Error: Status 0 returned from API request to https://www.reddit.com/api/request_promo.json?app=2x-client-production",
    "reduxInfo": "Redux Action Stack: \"FAILED_AD_FETCH\", \"RECEIVED_RECOMMENDED_SUBREDDITS\", \"SET_TITLE\", \"PLATFORM__SET_STATUS\", \"FETCHING_AD\", \"SAVE_SCROLL_POSITION\", {\"type\":\"PLATFORM__GOTO_PAGE_INDEX\",\"payload\":{\"pageIndex\":1,\"pathName\":\"/\",\"queryParams\":{\"utm_source\":\"mweb_redirect\",\"compact\":\"true\"},\"hashParams\":{},\"referrer\":\"\"}}, \"SET_TITLE\", \"PLATFORM__SET_STATUS\", \"FETCHING_RECOMMENDED_SUBREDDITS\"",
    "requestUrl": "/?utm_source=mweb_redirect&compact=true",
    "stack": "ResponseError: Status 0 returned from API request to https://www.reddit.com/api/request_promo.json?app=2x-client-production\n    at https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:10:13480\n    at https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:1:26912\n    at p.callback (https://www.redditstatic.com/mweb2x/ProductionClient.f0e9bc9b26ab29871b34.js:10:7098)\n    at p.crossDomainError (https://www.redditsta",
    "userAgent": "Mozilla/5.0 (Linux; Android 4.4.2; HTC Desire 620G dual sim Build/KOT49H) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.81 Mobile Safari/537.36"
}
```

</p></details>

### CODE REVIEW

File: src/lib/errorLog.js [hunk 1]
```diff
@@ -81,7 +81,10 @@ const parseError = error => {
     return {
       ...parseStackTrace(error.stack),
       message,
-      stack: error.stack,
+      // Truncating to prevent log lines from overflowing. The stack trace
+      // is a \n delimited string, and each line of the stack trace is ~120
+      // characters.
+      stack: error.stack.substring(0, 2048),
```
Comment (uzi) [inline src/lib/errorLog.js]: Should we do this... or should we do a split on '\n', take the first 10 elements and then rejoin them with '\n'?

Comment (wting) [inline src/lib/errorLog.js]: We care more about the character limit than the number of elements. For example, what if the first line was 4KB large?

That said, the *nicer* way is to split, count up to the limit, then rejoin.

Comment (prashtx) [inline src/lib/errorLog.js]: My two cents: If we're truncating log lines elsewhere based on character count, then this approach helps us factor the stack trace into that budget with more certainty.  It also results in less garbage generated for probably similar helpfulness to the developer.

Comment (schwers) [inline src/lib/errorLog.js]: There's another stack in `parseRejection` below, it might be better to do the truncation in `buildLogJSON`

**Comment by wting** [comment]
:haircut_woman:

File: src/lib/errorLog.js [hunk 2]
```diff
@@ -179,7 +176,8 @@ const buildLogJSON = details => {
     url,
     line,
     column,
-    stack,
+    // Prevent log lines from overflowing and becoming invalid JSON.
```
Comment (schwers) [inline src/lib/errorLog.js]: Nit: I really like how the first commit message in this PR makes it clear why 2048 was chosen, could you include the max line-length here too?

**Comment by schwers** [comment]
:fish:

File: src/lib/errorLog.js (hunk 2 above)
Comment (wting) [inline src/lib/errorLog.js]: Comment updated!

**Pull request merged**